osu.Game/Rulesets/Scoring/ScoreProcessor.cs

```diff
@@ -313,26 +336,29 @@ public double ComputeFinalLegacyScore(ScoringMode mode, ScoreInfo scoreInfo, int
             double accuracyRatio = scoreInfo.Accuracy;
             double comboRatio = maxAchievableCombo > 0 ? (double)scoreInfo.MaxCombo / maxAchievableCombo : 1;
 
+            extractScoringValues(scoreInfo.Statistics, out var current, out var maximum);
+
             // For legacy osu!mania scores, a full-GREAT score has 100% accuracy. If combined with a full-combo, the score becomes indistinguishable from a full-PERFECT score.
             // To get around this, the accuracy ratio is always recalculated based on the hit statistics rather than trusting the score.
             // Note: This cannot be applied universally to all legacy scores, as some rulesets (e.g. catch) group multiple judgements together.
-            if (scoreInfo.IsLegacyScore && scoreInfo.Ruleset.OnlineID == 3)
-            {
-                extractFromStatistics(
-                    ruleset,
-                    scoreInfo.Statistics,
-                    out double computedBaseScore,
-                    out double computedMaxBaseScore,
-                    out _,
-                    out _);
-
-                if (computedMaxBaseScore > 0)
-                    accuracyRatio = computedBaseScore / computedMaxBaseScore;
-            }
+            if (scoreInfo.IsLegacyScore && scoreInfo.Ruleset.OnlineID == 3 && maximum.BaseScore > 0)
+                accuracyRatio = current.BaseScore / maximum.BaseScore;
 
-            int computedBasicHitObjects = scoreInfo.Statistics.Where(kvp => kvp.Key.IsBasic()).Select(kvp => kvp.Value).Sum();
+            return ComputeScore(mode, accuracyRatio, comboRatio, current.BonusScore, maximum.HitObjects);
+        }
 
-            return ComputeScore(mode, accuracyRatio, comboRatio, getBonusScore(scoreInfo.Statistics), computedBasicHitObjects);
+        /// <summary>
+        /// Computes the total score from scoring values.
+        /// </summary>
+        /// <param name="mode">The <see cref="ScoringMode"/> to represent the score as.</param>
+        /// <param name="current">The current scoring values.</param>
+        /// <param name="maximum">The maximum scoring values.</param>
+        /// <returns>The total score computed from the given scoring values.</returns>
+        public double ComputeScore(ScoringMode mode, ScoringValues current, ScoringValues maximum)
```

Comment:
I kinda wish that this method could be static, but it can't because the `ComputeScore()` overload it calls also isn't (and can't be) static... Maybe could stick a `[Pure]` attribute on these two methods to convey that they don't change the `ScoreProcessor`'s state? Unfortunately those are just annotations and therefore it doesn't seem that any analyser actually checks that the method is actually pure, but it could be a way to let people know that it's safe to use them without worrying about polluting some other bit of state in the `ScoreProcessor`...

I guess an xmldoc mention could also suffice.

Reply:
Technically all of these methods (including `ExtractScoringValues`) could be static, and I originally contemplated moving them out to extensions. The only reason they aren't is because I'd have to:
1. Store `maxResult` used in `ExtractScoringValues`, likely to a `ConcurrentDictionary`?
2. Expose the accuracy/combo proportions and the classic multiplier, which are all currently either private or protected.

I agree it looks ugly though so I'm willing to consider it further.

Reply:
I think it's fine to leave be for now. The way my thinking goes, is that the steps to reach an "ideal world" structure here would be to:

* Split off `ScoreCalculator` that only has the actual math/data manipulation to be able to compute a total score from a `ScoreInfo`/whatever else (i.e. the methods that are candidates for being static)
* Break ruleset API so that rulesets create `ScoreCalculator`s instead of `ScoreProcessor`s, baking stuff like the accuracy/combo proportions and the classic multiplier into the `ScoreCalculator` that rulesets create (this is a pretty big break, unfortunately - not sure how it would be done in a backwards compatible manner)
* Gut all of the maths from `ScoreProcessor` and delegate the maths to `ScoreCalculator` instead, leaving `ScoreProcessor` as a pure state bookkeeping class that only responds to judgement applications/reverts in gameplay and updates state accordingly

The above is a lot and I'm not going to insist on this changing unless you feel like giving it a go. This PR is in the "not perfect but good enough" zone for me now.

Reply:
After reading through this PR I was going to suggest roughly the same thing as above (and that was before event reading this discussion thread, too).

Agree that the path forward is splitting the `Bindable` / gameplay portion of this class out, and also agree with the `ScoreCalculator` terminology to better reflect the stateless nature of the contained methods.

That said, I'm fine with the changes in this PR as an initial step in the right direction.